src/background/services/tabState.ts

```diff
@@ -0,0 +1,62 @@
+import { getHash } from '../utils'
+
+import { Tabs } from 'webextension-polyfill'
+import { WalletAddress } from '@interledger/open-payments'
+
+type State = {
+  lastPaymentTimestamp: number
+  expiresAtTimestamp: number
+}
+
+export class TabState {
+  private state = new WeakMap<Tabs.Tab, { [key: string]: State }>()
+
+  constructor() {}
+
+  private async getStateKey(
+    url: string,
+    walletAddress: WalletAddress
+  ): Promise<string> {
+    const hashUrl = await getHash(url)
+    const hashWalletAddress = await getHash(JSON.stringify(walletAddress))
```

Comment:
Maybe we can get rid of hashing given we're storing in memory only?

Reply:
Might be better to use `new WeakMap<Tabs.Tab, Map<string, State>>()` also.

Reply:
Also, do we need to keep entire `walletAddress`? Can't we only use `walletAddress.id`?

Reply:
I'd keep the hash  for consistency (in the feature we'll implement the exception list, that will also use the url, better have the hash everywhere)

Reply:
But exception list will be in storage (privacy part), and not updated on each payment. Speaking from performance perspective only - our extension will consume more battery.